test.js

```diff
@@ -1,304 +1,312 @@
-  fastify
-    .register(jwt, { secret: 'supersecret' })
-    .ready(function () {
+const test = require('tap').test
+const Fastify = require('fastify')
+
+const jwt = require('./jwt')
+
+test('register', function (t) {
+  t.plan(3)
+
+  t.test('expose jwt methods', function (t) {
+    t.plan(6)
+
+    const fastify = Fastify()
+    fastify.register(jwt, { secret: 'test' })
+
+    fastify.get('/methods', function (request, reply) {
+      t.ok(request.jwtVerify)
+      t.ok(reply.jwtSign)
+    })
+
+    fastify.ready(function () {
       t.ok(fastify.jwt.decode)
       t.ok(fastify.jwt.sign)
       t.ok(fastify.jwt.verify)
       t.ok(fastify.jwt.secret)
     })
-  fastify.get('/test', function (request, reply) {
-    t.ok(request.jwtVerify)
-    t.ok(reply.jwtSign)
-    reply.send({ foo: 'bar' })
-  })
-  fastify.listen(0, function (err) {
-    fastify.server.unref()
-    t.error(err)
-    rp({
-      method: 'GET',
-      uri: 'http://localhost:' + fastify.server.address().port + '/test',
-      json: true
-    }).then(function (response) {
-      t.ok(response)
-    }).catch(function (err) {
-      t.fail(err)
+
+    fastify.inject({
+      method: 'get',
+      url: '/methods'
     })
   })
-})
 
-test('fastify-jwt fails without secret', function (t) {
-  t.plan(1)
-  var fastify = Fastify()
-  fastify
-    .register(jwt)
-    .listen(0, function (err) {
-      t.is(err.message, 'missing secret')
-    })
-})
+  t.test('secret as a function', function (t) {
+    t.plan(2)
 
-test('sign and verify', function (t) {
-  t.plan(8)
-  var fastify = Fastify()
-  fastify.register(jwt, { secret: 'supersecret' })
+    const fastify = Fastify()
+    fastify.register(jwt, {
+      secret: function (request, reply, callback) {
+        callback(null, 'some-secret')
+      }
+    })
 
-  fastify.post('/signCallback', function (request, reply) {
-    reply.jwtSign(request.body.payload, function (err, token) {
-      return reply.send(err || { 'token': token })
+    fastify.post('/sign', function (request, reply) {
+      reply.jwtSign(request.body)
+        .then(function (token) {
+          return reply.send({ token })
+        })
     })
-  })
 
-  fastify.get('/verifyCallback', function (request, reply) {
-    request.jwtVerify(function (err, decoded) {
-      return reply.send(err || decoded)
+    fastify.get('/verify', function (request, reply) {
+      request.jwtVerify()
+        .then(function (decodedToken) {
+          return reply.send(decodedToken)
+        })
     })
-  })
 
-  fastify.post('/signPromise', function (request, reply) {
-    reply.jwtSign(request.body.payload).then(
-      function (token) {
-        return reply.send({ 'token': token })
-      }).catch(function (err) {
-        return reply.send(err)
+    fastify
+      .ready()
+      .then(function () {
+        fastify.inject({
+          method: 'post',
+          url: '/sign',
+          payload: { foo: 'bar' }
+        })
+          .then(function (signResponse) {
```

Comment:
Can you inline these `.then` chains? Something like this should work:
```javascript
fastify
  .ready()
  .then(function () {
    fastify.inject({
      method: 'get',
      ...
    }).then(function () {
      fastify.inject({
        method: 'get',
        ...
      }).then(function () {
        // do stuff
      })
  }).catch(function () {
    // I think you get the pattern :) 
  })
```

Reply:
Oh no not necessary to add .catch if they aren’t being used. I just added it to my example cause normally you’d have a catch in a .then promise chain. Great work on this refactor!